replaces path to .asar.unpacked if cmd includes .asar/node_modules

thisconnect opened the pull request:
js-ipfsd-ctl will not work inside an asar container, see https://github.com/ipfs/js-ipfsd-ctl/issues/94.

A possible solution is to exclude the ipfs binary from asar, with the unpackDir option

```bash
electron-packager ./ --asar.unpackDir=node_modules/go-ipfs-dep
```
(EDITED: changed `go-ipfs-dep/**` to `go-ipfs-dep` ) 

on MacOS: this will package everything expect go-ipfs-dep into `<appname>.asar`
and put go-ipfs-dep into `<appname>.asar.unpacked/node_modules/go-ipfs-dep`

This change request adds a check if `.asar/` is part of the path and changes it to `.asar.unpacked/`

Edit:

### How to test this change

Assuming you have a project using ipfsd-ctl, electron and electron-packager. 

Install https://github.com/thisconnect/js-ipfsd-ctl/tree/asar-unpacked
```
npm i thisconnect/js-ipfsd-ctl#asar-unpacked
```
build electron with `--asar.unpackDir=node_modules/go-ipfs-dep`
```
electron-packager ./ --asar.unpackDir=node_modules/go-ipfs-dep
```

Test if your app works and if everything (but go-ipfs-dep) is bundled in an .asar file

diasdavid commented:
@thisconnect I'm not sure if I understand all the implications here. If `go-ipfs` doesn't get bundled, then how would a standalone electron app that wants to use it work?

thisconnect commented:
`electron-packager ./ --asar.unpackDir=node_modules/go-ipfs-dep/**` does only exclude from the asar archive, but still bundle the files in the app. On MacOS I get
```
MyApp.app/Contents/Resources/app.asar
MyApp.app/Contents/Resources/app.asar.unpacked/node_modules/go-ipfs-dep
```

This PR checks if the path to the ipfs binary points to inside an asar archive and changes the path to asar.unpacked. It tests for `.asar/node_modules` which probably will not work on windows.

With this change It successfully bundle MyApp into an asar archive and am able to start the ipfsd-ctl deamon, that is in app.asar.unpacked. 

At this point I think it would be better to do this in the  findIpfsExecutable function https://github.com/thisconnect/js-ipfsd-ctl/blob/4453ea7caed94a00cac54d57c1e60d1e6420c1d7/src/node.js#L19

related docs https://electron.atom.io/docs/tutorial/application-packaging/#executing-binaries-inside-asar-archive

An other maybe better solution would be to be able to configure the path to the ipfs binary.

thisconnect commented:
The configuration could be done on as api or environment variable. As a ipfsd-ctl consumer I want to just use the api and **not** write my code differently for a packaging detail. 

I am unsure if configuring the path with an env variable is possible and a good idea :/

diasdavid commented:
Got it. I'm afraid I haven't used enough electron to know the full breadth of issues, but this PR doesn't seem to break our existing use cases so I'm fine trying it out :)

Could you add documentation of the issue and how to circumvent it with the packaging command you mention here so that others can try it too? 

@kyledrake, you've been hacking on electron and going through the same issues, mind sharing your thoughts here?

thisconnect commented:
Please do **NOT** merge this yet, it may not work on windows, due to forward slash. Not sure if that works with Node 4.x / npm 2.x when modules are not deduped.

### How to test this change

Assuming you have a project using ipfsd-ctl, electron and electron-packager. 

Install https://github.com/thisconnect/js-ipfsd-ctl/tree/asar-unpacked
```
npm i thisconnect/js-ipfsd-ctl#asar-unpacked
```
build electron with `--asar.unpackDir=node_modules/go-ipfs-dep/**`
```
electron-packager ./ --asar.unpackDir=node_modules/go-ipfs-dep/**
```

Test if your app works and if everything (but go-ipfs-dep) is bundled in an .asar file



I have not tried `electron-builder` yet, but maybe soon. With `electron-builder` asar is enabled by default, see https://github.com/electron-userland/electron-builder/wiki/Options#AsarOptions
also it seems `electron-builder` detects binaries and unpacks them automatically, though I am not sure how that works. I am going to try this soon.

thisconnect commented:
Ok tested `thisconnect/js-ipfsd-ctl#asar-unpacked` on windows 10.

for windows 
- added os.platform check to set the executable filename to `ipfs.exe`
- electron-packager wildcard (glob?) seems to work differently on windows, changed
`electron-packager ./ --asar.unpackDir=node_modules/go-ipfs-dep/**`
to
`electron-packager ./ --asar.unpackDir=node_modules/go-ipfs-dep/*`
so that it includes the dir and the binary

Please review https://github.com/ipfs/js-ipfsd-ctl/pull/155/files

diasdavid commented:
@thisconnect is this ready for review and merge?

thisconnect commented:
@diasdavid yes please. 

also please have a look at this line, https://github.com/ipfs/js-ipfsd-ctl/pull/155/files#diff-6ff379484cbabad48301d485db111c08R32 that should have been a different PR.

in the meantime the power port of my windows laptop melted. Do you use appveyor for windows CI?

diasdavid commented:
@thisconnect just enabled appveyor for this repo. Could you add the config file ([Tutorial](https://www.appveyor.com/docs/lang/nodejs-iojs/), [Great example](https://github.com/nodejs/nan/blob/master/appveyor.yml))

https://ci.appveyor.com/project/diasdavid/js-ipfsd-ctl

diasdavid commented:
Also, just added you to the JS team. You can now create branches on the main repo and PR directly from them. This helps collaboration between multiple people 

https://github.com/orgs/ipfs/teams/javascript-team/members 🌟

thisconnect commented:
Thanks @diasdavid I'll find time over the weekend and will create a branch for appveyor.

diasdavid commented:
Ok, I'll be more tranquil to merge this PR when I see the tests passing in Windows and in Linux CI :)

diasdavid commented on src/node.js:
```diff
@@ -19,7 +19,15 @@ const GRACE_PERIOD = 7500 // amount of ms to wait before sigkill
+  // electron-packager ./ --asar.unpackDir=node_modules/go-ipfs-dep
+  if (appRoot.includes(`.asar${path.sep}`)) {
+    appRoot = appRoot.replace(`.asar${path.sep}`, `.asar.unpacked${path.sep}`)
+  }
```
Could you add this to the README too? It is fine to leave it here, but I want to make sure it this note is searchable and linkable to

thisconnect replied:
Sure

thisconnect commented:
There is a failing test on windows https://ci.appveyor.com/project/diasdavid/js-ipfsd-ctl/build/21/job/ohabcvtp3jtlsgoa#L1610 ,
this to be the same issue as https://github.com/ipfs/js-ipfs-api/issues/408

diasdavid commented:
@thisconnect https://github.com/ipfs/js-ipfs-api/issues/408 shouldn't be an issue of this module specifically. Is there another test you can add to check that the module works?

thisconnect commented:
@diasdavid I have never built electron within CI, it should be possible but might take a long time. Not sure if it is worth right now.

I tried to add some info in the readme and a minimal example in examples/electron-asar, please have a look

thisconnect commented:
The "checks response of ipfs.util.addFromFs for windows" now tests too much. 

Do you prefer 

A) the deep equal that should cover everything about the response https://github.com/thisconnect/js-ipfsd-ctl/blob/46c422ebf1f5b11acecf81da74a553fbab4093d5/test/index.spec.js#L394-L406

B) or remove it and keep the rest https://github.com/thisconnect/js-ipfsd-ctl/blob/46c422ebf1f5b11acecf81da74a553fbab4093d5/test/index.spec.js#L408-L418

C) or keep it as it was before

diasdavid commented:
@thisconnect given that addFromFs has issues in Windows, I don't believe it to be a good test case to have. Better to skip (add a .skip) it for now.

diasdavid commented on test/index.spec.js:
```diff
@@ -390,10 +390,32 @@ describe('daemons', () => {
         if (err) throw err
```
see: https://github.com/ipfs/js-ipfsd-ctl/pull/155#issuecomment-326970190

diasdavid replied:
Here is an example: https://github.com/ipfs/npm-go-ipfs-dep/blob/master/src/index.js#L47-L60

thisconnect replied:
As it is a temporary skip, is it ok if I just `os.platform() === 'win32'` ?

https://nodejs.org/api/os.html#os_os_platform

thisconnect replied:
I don't get https://github.com/ipfs/npm-go-ipfs-dep/blob/master/src/index.js#L47-L60 sorry!! :( it also depends on another module (go-platform) and file https://github.com/ipfs/npm-go-ipfs-dep/blob/master/src/check-support.js

thisconnect commented:
Not sure what the last one is all about https://ci.appveyor.com/project/diasdavid/js-ipfsd-ctl/build/24/job/t4r5euk57pms9d20#L1610

diasdavid commented:
Seems like a go-ipfs thing, @Kubuxu ?

thisconnect commented:
https://github.com/thisconnect/js-ipfsd-ctl/blob/f3e49d54ecb143d130812dfc07ebd0788ae5aa72/test/index.spec.js#L438 checks for a repo.lock file this is in index.spec.js , something before did not close correctly?

Kubuxu commented:
Nothing changed in go-ipfs that could cause it. How does shutting down work? Does it wait for go-ipfs to shutdown? Is it station timeout or does it poll the process?

diasdavid commented:
> Nothing changed in go-ipfs that could cause it. How does shutting down work? 

It is the first time we are enabling Windows testing in this module, so really anything is worth ;)

Kubuxu commented:
It might be Windows, Windows is a bit neglected so it is possible that this is an issue.

diasdavid commented on test/index.spec.js:
```diff
@@ -431,35 +432,40 @@ describe('daemons', () => {
+          if (fs.existsSync(path.join(dir, 'api'))) {
+            cb(new Error('api file not removed'))
+          }
+          cb()
```
I didn't mean to skip the whole test, just to skip the assertion for repo.lock

diasdavid commented on test/index.spec.js:
```diff
@@ -431,35 +432,40 @@ describe('daemons', () => {
-          cb(new Error('api file not removed'))
+    // skip on windows
+    // https://github.com/ipfs/js-ipfsd-ctl/pull/155#issuecomment-326983530
+    if (isWindows) it.skip('start and stop')
```
add curly `{}`

thisconnect commented:
there was travis fail `Uncaught TypeError: Cannot read property 'constructor' of null` https://travis-ci.org/ipfs/js-ipfsd-ctl/builds/271713202?utm_source=github_status&utm_medium=notification

thisconnect commented:
api file assert has to go too https://ci.appveyor.com/project/diasdavid/js-ipfsd-ctl/build/26/job/gubdael6q0fl8fxu#L1685

diasdavid commented on test/index.spec.js:
```diff
@@ -404,11 +436,15 @@ describe('daemons', () => {
+          }
+          if (fs.existsSync(path.join(dir, 'api'))) {
+            cb(new Error('api file not removed'))
+          }
```
👍

thisconnect commented:
https://ci.appveyor.com/project/diasdavid/js-ipfsd-ctl-a9ywu/build/27/job/9ocd4i7lcflnxh6c#L129 :(
